### ISSUE TITLE
nix-shell behaves different than shell.nix

### ISSUE THREAD

**Issue by sengaya** [NixOS/nix#3273]
I would expect that this command
```
nix-shell -p "terraform.withPlugins (plugins: [ plugins.azurerm plugins.azuread plugins.helm plugins.null plugins.random plugins.kubernetes plugins.template ])"
```
does the same thing as this `shell.nix`:
```
{ pkgs ? import <nixpkgs> {} }:
with pkgs;
let
  terraform-aks = terraform.withPlugins (p: with p; [
    azurerm
    azuread
    helm
    null
    random
    kubernetes
    template
  ]);
in
mkShell {
  name = "Terraform-for-AKS";
  buildInputs = [ terraform-aks ];
}
```
The `nix-shell -p …` command works as expected. With the `shell.nix` all the plugins are still downloaded by terraform. Somehow in the latter scenario terraform does not detect the plugins properly, that's why I guess that the two environments differ.

**Comment by edolstra** [comment]
This seems like an issue with `mkShell`, which is defined in the nixpkgs repo. Could you reopen this issue there?

**Comment by FRidh** [comment]
cc @zimbatm 
Note this is more of a nixpkgs issue.

**Comment by zimbatm** [comment]
I am not able to reproduce the issue. The latter example works as expected.

**Comment by zimbatm** [comment]
`nix show-derivation $(nix-instantiate shell.nix)` doesn't show any other references

**Comment by sengaya** [comment]
After doing a `nix-channel --update` (I'm on macos) I can't reproduce it anymore. Not sure what changed to be honest. Thanks for your time and sorry for the noise.

Minor issue I found afterwards: The `shell.nix` that I provided won't work for the `null` plugin. It was missing. I had to change it to
```
  terraform-aks = terraform.withPlugins (p: [
     p.null
     ...
```
to make it work. If you consider this a bug, I can open a new ticket.
